src/user-pool.ts

```diff
@@ -190,44 +333,240 @@ export class UserPool {
   }
 
   /**
-   * Adds the `onX` trigger handlers.
+   * Creates an authentication challenge.
    *
-   * Example:
    * ```ts
-   * userPool.onCreateAuthChallenge(new Function(scope, "CreateAuthChallenge", async (event) => {
-   *   // do work and then:
+   * userPool.onCreateAuthChallenge(new Function(scope, id, async (event) => {
+   *   // handle event, set response properties, etc.
+   *   event.response... = ..
    *   return event;
    * }));
    * ```
    *
-   * @see UserPoolTriggerOnMethods
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-create-auth-challenge.html
+   * @default - no trigger configured
    */
-  static {
-    Object.keys(triggers).forEach(
-      (triggerName) =>
-        ((UserPool.prototype as any)[
-          `${triggerName.charAt(0).toUpperCase()}${triggerName.slice(1)}`
-        ] = function (handler: Function<any>) {
-          this.on(triggerName, handler);
-        })
-    );
+  public onCreateAuthChallenge(
+    handler: Trigger<lambda.CreateAuthChallengeTriggerEvent>
+  ): void {
+    return this.on("createAuthChallenge", handler);
   }
-}
 
-export type UserPoolTriggerOnMethods = Mixin<{
-  [triggerName in keyof UserPoolTriggers as triggerName extends `${infer Head}${infer Rest}`
-    ? `on${Uppercase<Head>}${Rest}`
-    : never]-?: (handler: UserPoolTriggers[triggerName]) => void;
-}>;
+  /**
+   * Amazon Cognito invokes this trigger to send email notifications to users.
+   *
+   * ```ts
+   * userPool.onCustomEmailSender(new Function(scope, id, async (event) => {
+   *   // handle event, set response properties, etc.
+   *   event.response... = ..
+   *   return event;
+   * }));
+   * ```
+   *
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-email-sender.html
+   * @default - no trigger configured
+   */
+  public onCustomEmailSender(
+    handler: Trigger<lambda.CustomEmailSenderTriggerEvent>
+  ): void {
+    return this.on("customEmailSender", handler);
+  }
 
-/**
- * Adds the `on*` User Pool triggers to the {@link UserPool}.
- *
- * ```ts
- * userPool.onCreateAuthChallenge(new Function(scope, "CreateAuthChallenge", async (event) => {
- *   // do work and then:
- *   return event;
- * }));
- * ```
- */
-export interface UserPool extends UserPoolTriggerOnMethods {}
+  /**
+   * A custom Message AWS Lambda trigger.
+   *
+   * ```ts
+   * userPool.onCustomMessage(new Function(scope, id, async (event) => {
+   *   // handle event, set response properties, etc.
+   *   event.response... = ..
+   *   return event;
+   * }));
+   * ```
+   *
+   * @see https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-message.html
+   * @default - no trigger configured
+   */
+  public onCustomMessage(
+    handler: Trigger<lambda.CustomMessageTriggerEvent>
+  ): void {
+    return this.on("customMessage", handler);
+  }
```

Comment:
Why did you need to get rid of the dynamic ones? Just curious, fine with explicit.

Reply:
I was worried it would confuse developers and we can't expand on doc and typedoc is probably terrible

Reply:
Although I was particularly proud of that. It felt very much like a macro lol

Reply:
Yeah, it was cool, but agree about explicit being a better experience.